resolwe_bio/processes/expression/star_quantification.py

```diff
@@ -0,0 +1,509 @@
+
+    return df
+
+
+def prepare_expression_set(rc, tpm, cpm, feature_dict, outfile_name):
+    """Prepare expression set output data."""
+    rc_exp = exp_to_df(rc, "RAW_COUNT")
+    tpm_exp = exp_to_df(tpm, "TPM")
+    cpm_exp = exp_to_df(cpm, "CPM")
+    rc_exp["GENE_SYMBOL"] = rc_exp["FEATURE_ID"].map(feature_dict)
+    input_features = rc_exp["FEATURE_ID"].tolist()
+    # Check if all of the input feature IDs could be mapped to the gene symbols
+    if not all(f_id in feature_dict for f_id in input_features):
+        print(
+            f"{sum(rc_exp.isnull().values.ravel())} feature(s) "
+            f"could not be mapped to the associated feature symbols."
+        )
+    # Merge with normalized expression values
+    exp_set = rc_exp.merge(tpm_exp, on="FEATURE_ID")
+    exp_set = exp_set.merge(cpm_exp, on="FEATURE_ID")
+    # Reorder columns
+    columns = ["FEATURE_ID", "GENE_SYMBOL", "RAW_COUNT", "TPM", "CPM"]
+    exp_set = exp_set[columns]
+    # Replace NaN values with empty string
+    exp_set.fillna("", inplace=True)
+
+    # Write to file
+    exp_set.to_csv(
+        outfile_name + ".txt.gz",
+        header=True,
+        index=False,
+        sep="\t",
+        compression="gzip",
+    )
+
+    # Write to JSON
+    df_dict = exp_set.set_index("FEATURE_ID").to_dict(orient="index")
+    with open(outfile_name + ".json", "w") as f:
+        json.dump({"genes": df_dict}, f, allow_nan=False)
+
+
+def expression_to_storage(rc_input, rc_output):
+    """Convert expressions file to JSON format."""
+
+    def isfloat(value):
+        """Check if value is float."""
+        try:
+            float(value)
+            return True
+        except ValueError:
+            return False
+
+    with io.TextIOWrapper(io.BufferedReader(gzip.open(rc_input))) as f:
+        # Split lines by tabs
+        # Ignore lines without a number in second column
+        # Build a dictionary of gene-expression pairs
+        exp = {
+            "genes": {
+                gene_exp[0]: float(gene_exp[1])
+                for gene_exp in (l.split("\t") for l in f)
+                if len(gene_exp) == 2 and isfloat(gene_exp[1])
+            }
+        }
+
+    with open(file=rc_output, mode="wt") as f:
+        json.dump(exp, f)
+
+    return rc_output
+
+
+class STARQuantification(ProcessBio):
+    """STAR quantification.
+
+    This process is based on the output of STAR aligner 'gene counts'.
+    Strandedness is detected with Salmon or it can be specified manually.
+    Finally, for normalization of gene counts TPM and CPM are used.
+    """
+
+    slug = "star-quantification"
+    name = "STAR gene quantification"
+    requirements = {
+        "expression-engine": "jinja",
+        "executor": {
+            "docker": {
+                "image": "public.ecr.aws/genialis/resolwebio/rnaseq:6.2.0",
+            },
+        },
+        "resources": {
+            "cores": 4,
+            "memory": 32768,
```

Comment:
Drop the default requirements to 2 cores and 16384 memory. More resources are needed in the case of large Salmon index files (containing a full genome decoy), but not with the smaller index files that are typically used.